src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -634,23 +634,37 @@ ModifyPartialQuerySupported(Query *queryTree, bool multiShardQuery,
 		foreach(targetEntryCell, queryTree->targetList)
 		{
 			TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+
+			/* skip resjunk entries: UPDATE adds some for ctid, etc. */
+			if (targetEntry->resjunk)
+			{
+				continue;
+			}
+
 			bool targetEntryPartitionColumn = false;
+			AttrNumber targetColumnAttrNumber = InvalidAttrNumber;
 
 			/* reference tables do not have partition column */
 			if (partitionColumn == NULL)
 			{
 				targetEntryPartitionColumn = false;
 			}
-			else if (targetEntry->resno == partitionColumn->varattno)
+			else
 			{
-				targetEntryPartitionColumn = true;
+				if (commandType == CMD_UPDATE)
+				{
+					if (targetEntry->resname)
+					{
+						targetColumnAttrNumber = get_attnum(resultRelationId,
```

Comment:
maybe add a single line comment explaining that it is not possible to give an alias to `UPDATE .. SET column `